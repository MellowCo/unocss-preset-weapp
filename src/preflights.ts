import type { Preflight, PreflightContext } from '@unocss/core'
import { entriesToCss, toArray } from '@unocss/core'
import type { Theme } from './theme'

const wxPrefix = ['page,root-portal-content,::before,::after']
const taroPrefix = ['*,::before,::after']
const uniappPrefix = ['uni-page-body,::before,::after']
// const defaultPrefix = ['*,::before,::after', '::backdrop']

export default function (isH5: boolean, platform: string): Preflight[] {
  return [
    {
      layer: 'preflights',
      getCSS(ctx: PreflightContext<Theme>) {
        if (ctx.theme.preflightBase) {
          const css = entriesToCss(Object.entries(ctx.theme.preflightBase))
          let preflightRoot = ctx.theme.preflightRoot

          if (!preflightRoot) {
            if (isH5)
              preflightRoot = platform === 'uniapp' ? uniappPrefix : taroPrefix
            else
              preflightRoot = wxPrefix
          }

          const roots = toArray(preflightRoot)
          return roots.map(root => `${root}{${css}}`).join('')
        }
      },
    },
  ]
}
