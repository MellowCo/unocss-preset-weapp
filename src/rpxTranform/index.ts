import type { UtilObject } from '@unocss/core'

export * from './taro'
export * from './uniapp'

export const pxRE = /-?[.\d]+px/g
export const rpxRE = /-?[.\d]+rpx/g
export const rpxOrPxRE = /-?[.\d]+r?px/g

export function cssRpxTransform(css: UtilObject, regex: RegExp, transform: (val: string) => string) {
  css.entries.forEach((i) => {
    let value = i[1]

    if (value && typeof value === 'string') {
      const matchs = Array.from(value.matchAll(regex))

      if (matchs.length !== 0) {
        matchs.forEach((m) => {
          const size = m[0]
          value = (value as string).replace(size, transform(size))
        })

        i[1] = value
      }
    }
  })
}
