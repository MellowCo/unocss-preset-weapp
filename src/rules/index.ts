/* @export-submodules */
export * from './align'
export * from './behaviors'
export * from './border'
export * from './color'
export * from './container'
export * from './decoration'
export * from './default'
export * from './flex'
export * from './gap'
export * from './grid'
export * from './layout'
export * from './position'
export * from './question-mark'
export * from './ring'
export * from './shadow'
export * from './size'
export * from './spacing'
export * from './static'
export * from './svg'
export * from './transform'
export * from './transition'
export * from './typography'
export * from './variables'

export * from './animation'
export * from './background'
export * from './safe-area'
export * from './line-clamp'
export * from './filters'
export * from './divide'
