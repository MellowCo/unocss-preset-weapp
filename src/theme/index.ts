/* @export-submodules */
export * from './animation'
export * from './colors'
export * from './default'
export * from './filters'
export * from './font'
export * from './misc'
export * from './preflight'
export * from './size'
export * from './types'
