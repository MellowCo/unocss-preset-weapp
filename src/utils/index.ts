export * from './mappings'
export * from './handlers'
export * from './variants'
export * from './utilities'
