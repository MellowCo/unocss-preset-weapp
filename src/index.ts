import type { Theme, ThemeAnimation } from './theme'
import { definePreset, type PresetOptions } from '@unocss/core'
import { extractorArbitraryVariants } from '@unocss/extractor-arbitrary-variants'
import { cacheTransformEscapESelector, defaultRules } from 'unplugin-transform-class/utils'
import preflights from './preflights'
import { taroCssPxTransform, taroH5CssRemTransform, uniAppVue2CssRpxTransform } from './rpxTranform'
import { rules } from './rules'
import { shorthands } from './shorthands'
import { theme } from './theme'

import { variants } from './variants'

export { colors, theme } from './theme'
export { parseColor } from './utils'

// support custom shadow color
export type { Theme, ThemeAnimation }

export interface DarkModeSelectors {
  /**
   * Selector for light variant.
   *
   * @default '.light'
   */
  light?: string

  /**
   * Selector for dark variant.
   *
   * @default '.dark'
   */
  dark?: string
}

export interface PresetWeappOptions extends PresetOptions {
  /**
   * Dark mode options
   *
   */
  dark?: 'class' | 'media' | DarkModeSelectors

  /**
   * @default false
   */
  attributifyPseudo?: boolean

  /**
   * Prefix for CSS variables.
   *
   * @default 'un-'
   */
  variablePrefix?: string

  /**
   * Utils prefix
   *
   * @default undefined
   */
  prefix?: string | string[]

  /**
   * Generate preflight
   *
   * @default true
   */
  preflight?: boolean

  /**
   * 是否转换微信class
   *
   * @default true
   */
  transform?: boolean

  /**
   * 平台
   * @default 'uniapp'
   */
  platform?: 'taro' | 'uniapp'

  /**
   * taro h5 rem 换算尺寸标准
   * @default 750
   * @link https://taro-docs.jd.com/taro/docs/size
   */
  designWidth?: number

  /**
   * taro 设计稿尺寸换算规则
   * @default '{ 640: 2.34 / 2, 750: 1, 828: 1.81 / 2}'
   * @link https://taro-docs.jd.com/taro/docs/size
   */
  deviceRatio?: Record<number, number>

  /**
   * taro webpack 版本
   * taro webpack4 和 webpack5 h5根字体大小，导致不同版本 rem 不一致
   * 见下面issues
   * @link https://github.com/NervJS/taro/issues/12361
   * @default webpack4
   */
  taroWebpack?: 'webpack4' | 'webpack5'

  /**
   * 是否为h5
   * @default false
   */
  isH5?: boolean

  /**
   * 自定义转换规则
   * @default https://github.com/MellowCo/unplugin-transform-class#options
   */
  transformRules?: Record<string, string>

  /**
   * wh 是否使用 rpx 为默认单位
   * @example
   * ```
   * whRpx: true
   * w-10 -> width: 10rpx
   * h-10 -> height: 10rpx
   * ```
   *
   * @example
   * ```
   * whRpx: false
   * w-10 -> width: 80rpx
   * h-10 -> height: 80rpx
   * ```
   * @default true
   */
  whRpx?: boolean

  /**
   * Enable arbitrary variants, for example `<div class="[&>*]:m-1 [&[open]]:p-2"></div>`.
   *
   * Disable this might slightly improve the performance.
   *
   * @default true
   */
  arbitraryVariants?: boolean
}

export const presetWeapp = definePreset((options: PresetWeappOptions = {}) => {
  options.dark = options.dark ?? 'class'
  options.attributifyPseudo = options.attributifyPseudo ?? false
  options.preflight = options.preflight ?? true
  options.variablePrefix = options.variablePrefix ?? 'un-'

  options.transform = options.transform ?? true
  options.isH5 = options.isH5 ?? false
  options.designWidth = options.designWidth ?? 750
  options.deviceRatio = options.deviceRatio ?? {
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2,
  }
  options.platform = options.platform ?? 'uniapp'
  options.taroWebpack = options.taroWebpack ?? 'webpack4'
  options.transformRules = options.transformRules ?? defaultRules
  options.whRpx = options.whRpx ?? true

  return {
    name: 'unocss-preset-weapp',
    theme: {
      ...theme,
      transformRules: options.transformRules,
      whRpx: options.whRpx,
    },
    rules,
    variants: variants(options),
    options,
    postprocess(css) {
      // 是否转义class
      if (options.transform)
        css.selector = cacheTransformEscapESelector(css.selector, options.transformRules)

      // 设置变量前缀
      if (options.variablePrefix !== 'un-')
        VarPrefixPostprocessor(options.variablePrefix!, css)

      // taro 处理 h5 和 小程序 px 和 rpx 转换
      if (options.platform === 'taro') {
        const { taroWebpack, designWidth, deviceRatio } = options

        if (options.isH5) {
          // h5 px rpx 转 rem
          taroH5CssRemTransform(css, taroWebpack!, designWidth!, deviceRatio!)
        }
        else {
          // 小程序 taro 处理 px 为 rpx
          taroCssPxTransform(css, designWidth!, deviceRatio!)
        }
      }

      // uniapp vue2 webpack: h5 rpx 处理
      if (options.platform === 'uniapp' && options.isH5)
        uniAppVue2CssRpxTransform(css)
    },
    preflights: options.preflight ? preflights(options.isH5, options.platform) : [],
    prefix: options.prefix,
    extractorDefault: options.arbitraryVariants === false
      ? undefined
      : extractorArbitraryVariants(),
    autocomplete: {
      shorthands,
    },
  }
})

function VarPrefixPostprocessor(prefix: string, obj: any) {
  obj.entries.forEach((i: any) => {
    i[0] = i[0].replace(/^--un-/, `--${prefix}`)
    if (typeof i[1] === 'string')
      i[1] = i[1].replace(/var\(--un-/g, `var(--${prefix}`)
  })
}

export default presetWeapp
