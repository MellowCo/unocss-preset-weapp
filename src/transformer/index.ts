import { defaultAttributes, defaultIgnoreNonValuedAttributes } from 'unplugin-attributify-to-class/utils'
import { defaultRules } from 'unplugin-transform-class/utils'
import transformerClass from './transformerClass'
import { extractorAttributify, transformerAttributify } from './transformerAttributify'

export {
  transformerClass,
  extractorAttributify,
  transformerAttributify,
  defaultAttributes,
  defaultIgnoreNonValuedAttributes,
  defaultRules,
}
