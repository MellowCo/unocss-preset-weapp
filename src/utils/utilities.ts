import type { CSSEntries, CSSObject, DynamicMatcher, ParsedColorValue, RuleContext, StaticRule, VariantContext } from '@unocss/core'
import { isString, toArray } from '@unocss/core'
import { cacheRestoreSelector } from 'unplugin-transform-class/utils'
import { getStringComponent, getStringComponents } from '@unocss/rule-utils'
import type { Theme } from '../theme'
import { colorOpacityToString, colorToString, parseCssColor } from './colors'

import { handler as h } from './handlers'
import { cssMathFnRE, directionMap, globalKeywords } from './mappings'
import { bracketTypeRe, numberWithUnitRE } from './handlers/regex'

export const CONTROL_MINI_NO_NEGATIVE = '$$mini-no-negative'

/**
 * Provide {@link DynamicMatcher} function returning spacing definition. See spacing rules.
 * @param {string} propertyPrefix - Property for the css value to be created. Postfix will be appended according to direction matched.
 * @see {@link directionMap}
 */
export function directionSize(propertyPrefix: string): DynamicMatcher {
  return ([_, direction, size]: string[], { theme }: RuleContext<Theme>): CSSEntries | undefined => {
    size = cacheRestoreSelector(size, theme?.transformRules)
    const v = theme.spacing?.[size || 'DEFAULT'] ?? h.bracket.cssvar.global.auto.fraction.remToRpx(size)
    if (v != null)
      return directionMap[direction].map(i => [`${propertyPrefix}${i}`, v])
  }
}

type ThemeColorKeys = 'colors' | 'borderColor' | 'backgroundColor' | 'textColor' | 'shadowColor' | 'accentColor'

function getThemeColorForKey(theme: Theme, colors: string[], key: ThemeColorKeys = 'colors') {
  let obj = theme[key] as Theme['colors'] | string
  let index = -1

  for (const c of colors) {
    index += 1
    if (obj && typeof obj !== 'string') {
      const camel = colors.slice(index).join('-').replace(/(-[a-z])/g, n => n.slice(1).toUpperCase())
      if (obj[camel])
        return obj[camel]

      if (obj[c]) {
        obj = obj[c]
        continue
      }
    }
    return undefined
  }

  return obj
}

/**
 * Obtain color from theme by camel-casing colors.
 * @param theme
 * @param colors
 * @param key
 */
function getThemeColor(theme: Theme, colors: string[], key?: ThemeColorKeys) {
  return getThemeColorForKey(theme, colors, key) || getThemeColorForKey(theme, colors, 'colors')
}

/**
 * Split utility shorthand delimited by / or :
 * @param body
 * @param type
 * @param theme
 */
export function splitShorthand(body: string, type: string, theme: Theme) {
  const [front, rest] = getStringComponent(cacheRestoreSelector(body, theme.transformRules), '[', ']', ['/', ':', '_']) ?? []

  if (front != null) {
    const match = (front.match(bracketTypeRe) ?? [])[1]

    if (match == null || match === type)
      return [front, rest]
  }
}

/**
 * Parse color string into {@link ParsedColorValue} (if possible). Color value will first be matched to theme object before parsing.
 * See also color.tests.ts for more examples.
 *
 * @example Parseable strings:
 * 'red' // From theme, if 'red' is available
 * @param key
 * 'red-100' // From theme, plus scale
 * 'red-100_20' // From theme, plus scale/opacity
 * '[rgb(100,2,3)]/[var(--op)]' // Bracket with rgb color and bracket with opacity
 *
 * @param {string} body - Color string to be parsed.
 * @param {Theme} theme - {@link Theme} object.
 * @return {ParsedColorValue|undefined}  {@link ParsedColorValue} object if string is parseable.
 */
export function parseColor(body: string, theme: Theme, key?: ThemeColorKeys): ParsedColorValue | undefined {
  const split = splitShorthand(body, 'color', theme)
  if (!split)
    return

  const [main, opacity] = split

  const colors = main
    .replace(/([a-z])([0-9])/g, '$1-$2')
    .split(/-/g)
  const [name] = colors

  if (!name)
    return

  let color: string | undefined
  const bracket = h.bracketOfColor(main)
  const bracketOrMain = bracket || main

  if (h.numberWithUnit(bracketOrMain))
    return

  if (/^#[\da-fA-F]+$/.test(bracketOrMain))
    color = bracketOrMain
  else if (/^hex-[\da-fA-F]+$/.test(bracketOrMain))
    color = `#${bracketOrMain.slice(4)}`
  else if (main.startsWith('$'))
    color = h.cssvar(main)

  color = color || bracket

  if (!color) {
    const colorData = getThemeColor(theme, [main], key)
    if (typeof colorData === 'string')
      color = colorData
  }

  let no = 'DEFAULT'
  if (!color) {
    let colorData
    const [scale] = colors.slice(-1)
    if (scale.match(/^\d+$/)) {
      no = scale
      colorData = getThemeColor(theme, colors.slice(0, -1))
      if (!colorData || typeof colorData === 'string')
        color = undefined
      else
        color = colorData[no] as string
    }
    else {
      colorData = getThemeColor(theme, colors, key)
      if (!colorData && colors.length <= 2) {
        [, no = no] = colors
        colorData = getThemeColor(theme, colors, key)
      }

      if (typeof colorData === 'string')
        color = colorData
      else if (no && colorData)
        color = colorData[no] as string
    }
  }

  return {
    opacity,
    name,
    no,
    color,
    cssColor: parseCssColor(color),
    alpha: h.bracket.cssvar.percent(opacity ?? ''),
  }
}

/**
 * Provide {@link DynamicMatcher} function to produce color value matched from rule.
 *
 * @see {@link parseColor}
 *
 * @example Resolving 'red' from theme:
 * colorResolver('background-color', 'background')('', 'red')
 * return { 'background-color': '#f12' }
 *
 * @example Resolving 'red-100' from theme:
 * colorResolver('background-color', 'background')('', 'red-100')
 * return { '--un-background-opacity': '1', 'background-color': 'rgba(254,226,226,var(--un-background-opacity))' }
 *
 * @example Resolving 'red-100_20' from theme:
 * colorResolver('background-color', 'background')('', 'red-100/20')
 * return { 'background-color': 'rgba(204,251,241,0.22)' }
 *
 * @example Resolving 'hex-124':
 * colorResolver('color', 'text')('', 'hex-124')
 * return { '--un-text-opacity': '1', 'color': 'rgba(17,34,68,var(--un-text-opacity))' }
 *
 * @param {string} property - Property for the css value to be created.
 * @param {string} varName - Base name for the opacity variable.
 * @param key
 * @param {Function} [shouldPass] - Function to decide whether to pass the css.
 * @return {@link DynamicMatcher} object.
 */
export function colorResolver(property: string, varName: string, key?: ThemeColorKeys, shouldPass?: (css: CSSObject) => boolean): DynamicMatcher {
  return ([, body]: string[], { theme }: RuleContext<Theme>): CSSObject | undefined => {
    const data = parseColor(body, theme, key)

    if (!data)
      return

    const { alpha, color, cssColor } = data

    const css: CSSObject = {}
    if (cssColor) {
      if (alpha != null) {
        css[property] = colorToString(cssColor, alpha)
      }
      else {
        const opacityVar = `--un-${varName}-opacity`
        const result = colorToString(cssColor, `var(${opacityVar})`)
        if (result.includes(opacityVar))
          css[opacityVar] = colorOpacityToString(cssColor)
        css[property] = result
      }
    }
    else if (color) {
      if (alpha != null) {
        css[property] = colorToString(color, alpha)
      }
      else {
        const opacityVar = `--un-${varName}-opacity`
        const result = colorToString(color, `var(${opacityVar})`)
        if (result.includes(opacityVar))
          css[opacityVar] = 1
        css[property] = result
      }
    }

    if (shouldPass?.(css) !== false)
      return css
  }
}

export function colorableShadows(shadows: string | string[], colorVar: string) {
  const colored = []
  shadows = toArray(shadows)
  for (let i = 0; i < shadows.length; i++) {
    // shadow values are between 3 to 6 terms including color
    const components = getStringComponents(shadows[i], ' ', 6)
    if (!components || components.length < 3)
      return shadows

    let isInset = false
    const pos = components.indexOf('inset')
    if (pos !== -1) {
      components.splice(pos, 1)
      isInset = true
    }

    let colorVarValue = ''
    if (parseCssColor(components.at(0))) {
      const color = parseCssColor(components.shift())
      if (color)
        colorVarValue = `, ${colorToString(color)}`
    }
    else if (parseCssColor(components.at(-1))) {
      const color = parseCssColor(components.pop())
      if (color)
        colorVarValue = `, ${colorToString(color)}`
    }

    colored.push(`${isInset ? 'inset ' : ''}${components.join(' ')} var(${colorVar}${colorVarValue})`)
  }

  return colored
}

export function hasParseableColor(color: string | undefined, theme: Theme, key: ThemeColorKeys) {
  return color != null && !!parseColor(color, theme, key)?.color
}

export function resolveBreakpoints({ theme, generator }: Readonly<VariantContext<Theme>>, key: 'breakpoints' | 'verticalBreakpoints' = 'breakpoints') {
  let breakpoints: Record<string, string> | undefined
  if (generator.userConfig && generator.userConfig.theme)
    breakpoints = (generator.userConfig.theme as any)[key]

  if (!breakpoints)
    breakpoints = theme[key]

  return breakpoints
    ? Object.entries(breakpoints)
      .sort((a, b) => Number.parseInt(a[1].replace(/[a-z]+/gi, '')) - Number.parseInt(b[1].replace(/[a-z]+/gi, '')))
      .map(([point, size]) => ({ point, size }))
    : undefined
}

export function resolveVerticalBreakpoints(context: Readonly<VariantContext<Theme>>) {
  return resolveBreakpoints(context, 'verticalBreakpoints')
}

export function makeGlobalStaticRules(prefix: string, property?: string): StaticRule[] {
  return globalKeywords.map(keyword => [`${prefix}-${keyword}`, { [property ?? prefix]: keyword }])
}

export function getBracket(str: string, open: string, close: string) {
  if (str === '')
    return

  const l = str.length
  let parenthesis = 0
  let opened = false
  let openAt = 0
  for (let i = 0; i < l; i++) {
    switch (str[i]) {
      case open:
        if (!opened) {
          opened = true
          openAt = i
        }
        parenthesis++
        break

      case close:
        --parenthesis
        if (parenthesis < 0)
          return

        if (parenthesis === 0) {
          return [
            str.slice(openAt, i + 1),
            str.slice(i + 1),
            str.slice(0, openAt),
          ]
        }
        break
    }
  }
}

export function getComponent(str: string, open: string, close: string, separators: string | string[]) {
  if (str === '')
    return

  if (isString(separators))
    separators = [separators]

  if (separators.length === 0)
    return

  const l = str.length
  let parenthesis = 0
  for (let i = 0; i < l; i++) {
    switch (str[i]) {
      case open:
        parenthesis++
        break

      case close:
        if (--parenthesis < 0)
          return
        break

      default:
        for (const separator of separators) {
          const separatorLength = separator.length
          if (separatorLength && separator === str.slice(i, i + separatorLength) && parenthesis === 0) {
            if (i === 0 || i === l - separatorLength)
              return
            return [
              str.slice(0, i),
              str.slice(i + separatorLength),
            ]
          }
        }
    }
  }

  return [
    str,
    '',
  ]
}

export function getComponents(str: string, separators: string | string[], limit?: number) {
  limit = limit ?? 10
  const components = []
  let i = 0
  while (str !== '') {
    if (++i > limit)
      return
    const componentPair = getComponent(str, '(', ')', separators)
    if (!componentPair)
      return
    const [component, rest] = componentPair
    components.push(component)
    str = rest
  }
  if (components.length > 0)
    return components
}

export function isCSSMathFn(value: string | undefined) {
  return value != null && cssMathFnRE.test(value)
}

export function isSize(str: string) {
  if (str[0] === '[' && str.slice(-1) === ']')
    str = str.slice(1, -1)
  return cssMathFnRE.test(str) || numberWithUnitRE.test(str)
}
