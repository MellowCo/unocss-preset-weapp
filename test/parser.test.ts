import { createGenerator } from '@unocss/core'
import { expect, it } from 'vitest'
import presetWeapp from '../src/index'
import { variantMatcher } from '../src/utils'

it('split string with custom separator', async () => {
  const uno = await createGenerator({
    presets: [
      presetWeapp(),
    ],
    separators: ['__', '_a_'],
  })

  const cssTargets = [
    'hover__bg-green-500',
    'active__bg-red-500',
    'hover_a_bg-green-500',
    'active_a_bg-red-500',
  ]

  const { css } = await uno.generate(cssTargets.join(' '), { preflights: false })
  expect(css).toMatchSnapshot()
})

it('unable to generate token variant with explicit separator without dash', async () => {
  const uno = await createGenerator({
    presets: [
      presetWeapp(),
    ],
    separators: '_c_',
  })

  const cssTargets = [
    'hover_c_bg-green-500',
    'active_c_bg-red-500',
  ]

  const { css } = await uno.generate(cssTargets.join(' '), { preflights: false })
  expect(css).toMatchSnapshot()
})

it('multiple handlers for single variant', async () => {
  const uno = await createGenerator({
    rules: [
      ['foo', { name: 'bar' }],
    ],
    variants: [
      variantMatcher('pre', [
        () => ({ prefix: '.prefix1 ' }),
        () => ({ prefix: '.prefix2 ' }),
      ]),
    ],
  })

  const { css } = await uno.generate([
    'pre:foo',
  ].join(' '), { preflights: false })

  expect(css).toMatchSnapshot()
})
