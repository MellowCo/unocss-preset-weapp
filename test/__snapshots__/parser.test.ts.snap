// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`split string with custom separator 1`] = `
"/* layer: default */
.hover__bg-green-500:hover{--un-bg-opacity:1;background-color:rgb(34 197 94 / var(--un-bg-opacity));}
.hover_a_bg-green-500:hover{--un-bg-opacity:1;background-color:rgb(34 197 94 / var(--un-bg-opacity));}
.active__bg-red-500:active{--un-bg-opacity:1;background-color:rgb(239 68 68 / var(--un-bg-opacity));}
.active_a_bg-red-500:active{--un-bg-opacity:1;background-color:rgb(239 68 68 / var(--un-bg-opacity));}"
`;

exports[`unable to generate token variant with explicit separator without dash 1`] = `
"/* layer: default */
.hover_c_bg-green-500:hover{--un-bg-opacity:1;background-color:rgb(34 197 94 / var(--un-bg-opacity));}
.active_c_bg-red-500:active{--un-bg-opacity:1;background-color:rgb(239 68 68 / var(--un-bg-opacity));}"
`;
